File: Python_Solutions/Partition_List.py
"""
Leetcode Challenge: Partition List

Given the head of a linked list and a value x, 
partition it such that all nodes less than x come before nodes 
greater than or equal to x.

You should preserve the original relative order of the nodes 
in each of the two partitions.

8/12/24
Mark Robuck
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    """
    Partition a linked list based on less than and greater than a value.
    Args:
        head(ListNode): original linked list
        x(int): value to partition linked list based on
    Returns:
        ListNode: partitioned linked list
    """
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        #create dummy nodes for each partition
        less_dummy = ListNode(0)
        greater_dummy = ListNode(0)

        #pointers for current end of each partition
        less = less_dummy
        greater = greater_dummy

        #iterate through linked list
        current = head
        while current:
            #append to nodes less than x
            if current.val < x:
                less.next = current
                less = less.next
            #append to nodes greater than x
            else:
                greater.next = current
                greater = greater.next
            current = current.next

        #set end of greater partition and combine partitions
        greater.next = None
        less.next = greater_dummy.next

        return less_dummy.next

File: Java_Solutions/Partition_List.java
/*
* Leetcode Challenge: Partition List
*
* Given the head of a linked list and a value x, 
* partition it such that all nodes less than x come before nodes 
* greater than or equal to x.
*
*You should preserve the original relative order of the nodes 
*in each of the two partitions.
*
* 8/12/24
* Mark Robuck
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    /*
    * Partition a linked list based on less than and greater than a value.
    * @params head, original linked list
    * @params x, value to partition linked list based on
    * @return partitioned linked list
    */
    public ListNode partition(ListNode head, int x) {
        //create dummy nodes for each partition
        ListNode lessDummy = new ListNode(0);
        ListNode greaterDummy = new ListNode(0);

        //pointers for current end of each partition
        ListNode less = lessDummy;
        ListNode greater = greaterDummy;

        //iterate through linked list
        ListNode current = head;
        while (current != null) {
            //append to nodes less than x
            if (current.val < x) {
                less.next = current;
                less = less.next;
            } else {
            //append to nodes greater than x
                greater.next = current;
                greater = greater.next;
            }
            current = current.next;
        }

        //set end of greater partition and combine partitions
        greater.next = null;
        less.next = greaterDummy.next;

        return lessDummy.next;
        
    }
}
